xbout/load.py

```diff
@@ -151,31 +155,82 @@ def _arrange_for_concatenation(filepaths, nxpe=1, nype=1):
     return paths_grid, concat_dims
 
 
-def _trim(ds, ghosts={}, keep_guards=True):
+def _trim(ds, ghosts, guards={}, keep_guards={}, nxpe=1, nype=1):
     """
     Trims all ghost and guard cells off a single dataset read from a single
     BOUT dump file, to prepare for concatenation.
 
     Parameters
     ----------
     ghosts : dict, optional
+        Number of ghost cells along each dimension, e.g. {'x': 2, 't': 0}
     guards : dict, optional
+        Number of guard cells along each dimension, e.g. {'x': 2, 'y': 2}
     keep_guards : dict, optional
+        Whether or not to preserve the guard cells along each dimension, e.g.
+        {'x': True, 'y': False}
     """
 
-    # TODO generalise this function to handle guard cells being optional
-    if not keep_guards:
-        raise NotImplementedError
+    if any(keep_guards.values()):
+        # Work out if this particular dataset contains any guard cells
+        # Relies on a change to xarray so datasets always have source encoding
+        # See xarray GH issue #2550
+        lower_guards, upper_guards = _infer_contains_guards(
+            ds.encoding['source'], nxpe, nype)
+    else:
+        lower_guards, upper_guards = {}, {}
 
     selection = {}
     for dim in ds.dims:
-        if ghosts.get(dim, False):
-            selection[dim] = slice(ghosts[dim], -ghosts[dim])
+        # Check for guard cells, otherwise use ghost cells, else leave alone
+        if keep_guards.get(dim, False):
+            if lower_guards.get(dim, False):
+                lower = None
+            else:
+                lower = max(ghosts[dim], guards[dim])
+        elif ghosts.get(dim, False):
+            lower = ghosts[dim]
+        else:
+            lower = None
+        if keep_guards.get(dim, False):
+            if upper_guards.get(dim, False):
+                upper = None
+            else:
+                upper = -max(ghosts[dim], guards[dim])
+        elif ghosts.get(dim, False):
+            upper = -ghosts[dim]
+        else:
+            upper = None
```

Comment:
The logic for `upper` is almost the same as for `lower` -- is it possible to pull out into a helper function?

Reply:
Probably - I'll have a go

Reply:
As they are not completely identical, it might be more trouble that its worth

Reply:
I managed to factor this out, so do you think this is ready to merge now @ZedThree ?